contracts/eventRegister/PhasedEventRegister.sol

```diff
@@ -0,0 +1,182 @@
+        __Context_init();
+        _grantRole(EVENT_MANAGER_ROLE, _msgSender());
+        _transferOwnership(_msgSender());
+    }
+
+    /**
+     * @notice Constructor for compatibility. Grants admin role to deployer.
+     */
+    constructor() {
+        _grantRole(DEFAULT_ADMIN_ROLE, _msgSender());
+    }
+
+    /**
+     * @notice Grants the EVENT_MANAGER_ROLE to a specified account.
+     * @param account The address to be granted the EVENT_MANAGER_ROLE.
+     */
+    function grantEventManagerRole(address account) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        grantRole(EVENT_MANAGER_ROLE, account);
+    }
+
+    /**
+     * @notice Revokes the EVENT_MANAGER_ROLE from a specified account.
+     * @param account The address from which the EVENT_MANAGER_ROLE will be revoked.
+     */
+    function revokeEventManagerRole(address account) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        revokeRole(EVENT_MANAGER_ROLE, account);
+    }
+
+    /**
+     * @notice Creates a new event with the given name and number of phases.
+     * @param _name The name of the event to be created.
+     * @param _totalPhases The number of registration phases.
+     *
+     * Requirements:
+     * - Event IDs start from 1.
+     * - All phase IDs for this event will be [1.._totalPhases].
+     */
+    function createEvent(string memory _name, uint256 _totalPhases) external onlyRole(EVENT_MANAGER_ROLE) {
+        nextEventId++;
+        uint256 eventId = nextEventId;
+
+        events[eventId] = Event({ id: eventId, name: _name, totalPhases: _totalPhases });
+
+        for (uint256 i = 1; i <= _totalPhases; i++) {
+            phaseOpen[eventId][i] = true;
+            emit PhaseOpened(eventId, i);
+        }
+
+        emit EventCreated(eventId, _name);
+    }
+
+    /**
+     * @notice Opens registrations for a specific phase of an event.
+     * @param _eventId The ID of the event.
+     * @param _phaseId The phase to be opened.
+     */
+    function openPhase(uint256 _eventId, uint256 _phaseId) external onlyRole(EVENT_MANAGER_ROLE) {
+        _validateIds(_eventId, _phaseId);
+        phaseOpen[_eventId][_phaseId] = true;
+        emit PhaseOpened(_eventId, _phaseId);
+    }
+
+    /**
+     * @notice Closes registrations for a specific phase of an event.
+     * @param _eventId The ID of the event.
+     * @param _phaseId The phase to be closed.
+     */
+    function closePhase(uint256 _eventId, uint256 _phaseId) external onlyRole(EVENT_MANAGER_ROLE) {
+        _validateIds(_eventId, _phaseId);
+        phaseOpen[_eventId][_phaseId] = false;
+        emit PhaseClosed(_eventId, _phaseId);
+    }
+
+    /**
+     * @notice Registers the caller for a specific phase of an event.
+     * @param _eventId The ID of the event.
+     * @param _phaseId The phase to register for.
+     */
+    function register(uint256 _eventId, uint256 _phaseId) external {
+        _validateIds(_eventId, _phaseId);
+        require(phaseOpen[_eventId][_phaseId], "Phase closed");
```

Comment:
Use defined errors and if/revert rather than require, as seen here:
https://github.com/taikoxyz/trailblazer-contracts/blob/0c5c1dc8f66fb1d68e7c05db43d62d2a70527d66/contracts/trailblazers-season-4/TrailblazersS1BadgesV9.sol#L42